### ISSUE TITLE
Error handling / how functions return strings

### ISSUE THREAD

**Issue by nyurik** [gquintard/varnish-rs#109]
All string-returning Varnish functions must store strings in a workspace. A macro-generated wrapper does this for String/&str/CString/&CStr values. This results in several failure modes that we might want to rethink. The main issue is that some failures occur **after** the Rust function exits, thus developer has no way to deal with the errors.

### String contains `\0`
Rust strings may contain NULL characters. The wrapper can either panic, abort the task (responds with HTTP 500?), or truncate the string, dropping anything after the NULL character.   Affected return types:  `String`, `&str`

### String is too big to fit into workspace
Running `WS_Alloc(ws, size)` C function may result in NULL, i.e. out of memory. This might be due to the string being too big, or other issues. The only options here is to panic or abort the task.  This affects all string types like `String`, `&str`, `CString`, and `&CStr`, with the only exception being `VCL_STRING` because it has been created by the Rust developer in a workspace already.  Note that this also affects IP addresses and probes.

### Incorrect `VCL_STRING` creation
`VCL_STRING` offers a way to handle the above two problems in the Rust code because vmod author can copy content into the workspace directly. The issue is that this struct is defined as `pub struct VCL_STRING(pub *const ::std::ffi::c_char)`  in `ffi` module, so it can be created incorrectly without the `unsafe` keyword:

```rust
fn foo() -> VCL_STRING {
    let s = CString::from(c"hello");
    VCL_STRING(s.as_ptr())  // s will be dropped on exit - dangling pointer
}
```

In other words, we do not want `VCL_STRING` to be instantiated in a `safe` code. Unfortunately, I cannot make bindgen to generate `VCL_STRING(pub(crate) *c_char)`  - it always creates `pub`, so I may need to do a workaround - bindgen will create `PUB_VCL_STRING` (or some other name), and I will implement another `VCL_STRING` type with unsafe constructor.

### vmod-level attribute settings
We can introduce a new parameter on the `#[varnish::vmod]` attribute with some sensible defaults:
* `on_oom=panic|abort_task` - action if `WS_Alloc(ws, size)` returns `NULL`.  Defaults to `abort_task`
* `on_null_in_str=panic|abort_task|trim` -  action if a Rust function returns a string with `NULL`. Not certain of the default.

### Encourage Workspace usage

We could encourage users to work directly with the workspace -- which will return `VCL_STRING` - returnable to Varnish.  This way it will be user's responsibility to convert their strings to `&CStr`, and thus deal with both OOM and NULL characters.

```rust
fn get_string_or_err(ws: &mut Workspace) -> Result<VCL_STRING, VclError> {
  // store fn returns a safe value to return, or will fail if OOM
  ws.store_c_str(c"something")
}

fn get_string(ws: &mut Workspace) -> VCL_STRING {
  // same as above, but the user can handle errors (in this case - panic on OOM)
  ws.store_c_str(c"something").unwrap()
}
```

**Comment by nyurik** [comment]
P.S. I submitted a PR for bindgen to allow custom accessibility generation. See https://github.com/rust-lang/rust-bindgen/pull/2967

**Comment by gquintard** [comment]
Hi,

Yes, it's somewhat related, but we now have functions like this: https://docs.rs/varnish/latest/varnish/vcl/struct.Workspace.html#method.copy_bytes_with_null which should be enough for what you need.
Note that there's no guarantee your request message will be a C string, so you will have a discrepancy between how loose C is and how strict rust is. You might be better using `.copy_blob()` instead, depending on what you intend to do with the string.

Another, more space efficient would be to use the String directly, add a`\0` at the end and stash it into a `VPriv` and just return a pointer to the String, but that only works if you promise to keep the String around until the transaction is done

**Comment by gquintard** [comment]
correct, if you want to return a `VCL_STRING` (a `char *` , really), you have to know what you are doing and guarantee a fee things. If you want safe code, you have to accept that we are going to have a few copies/inefficiencies.

**Comment by gquintard** [comment]
@nyurik , we could have a new type that equates to `VCL_STRING` but that is exclusively produces by `ws`, so that we know it's safe to return?

**Comment by nyurik** [comment]
@gquintard i think that would be the best approach, yes.  Want to hack on it?  should be fairly easy to create

**Comment by gquintard** [comment]
as time allows, yes, but I'm happy if someone else wants to steal it from me :-)

**Comment by nyurik** [comment]
* `WsStr` seems the most fitting - its a reference rather than owned value, and is stored in WS
* `WsStr<'ws>` is the proper type because its lifetime should be tied to the WS
* `WsBlob<'ws>` is the other type that reflects `VCL_BLOB`
* I think `fn finish(mut self) -> WsStr<'ws>` is a better sig
* `WsStr::as_vcl_string(&self)` would return `VCL_STRING`. The method is safe, as long as it is not de-referenced
* `WsStr::as_cstr(&self)` would return `CStr`
* A user function that returns `VCL_STRING` must be unsafe, whereas it can safely return `WsStr`

**Comment by gquintard** [comment]
And yeah, `WsStr` works too

**Comment by gquintard** [comment]
>  how will Varnish handle that "string"? 5xx? UB?

You are thinking at a way too high level here :-) The VCL API is "I'm returning a `char *` that is null-terminated and that you can count on until the VCL transaction is done". Other restrictions on the strings depends on where that `VCL_STRING` is used (body, header, query string, etc.) and you have no way of knowing/enforcing it.

> ```
> // should the old method be considered `unsafe` also?
> /// Finish writing to the [`WsBuffer`], returning the allocated [`VCL_STRING`].
> pub fn finish(mut self) -> VCL_STRING
> ```
I would side with the way `rust` handles it: generating a pointer is safe, consuming it isn't. So I think it makes sense that the old function is safe, but returning it from a vmod isn't, but I agree it's a bit weird.

**Comment by gquintard** [comment]
Ah, sorry, I missed that from Yuri.

I agree with you:
- one function is safe and checks the content, and returns an error if that fails
- the other is unsafe and just YOLOs the pointer without checking the content

**Comment by gquintard** [comment]
that would mean creating two more `Ws*Buffer`, which can work, but isn't that becoming unwieldy?

At a minimum, for the `unsafe` version, we can just use `WsTempBuffer`, make the suffix `\0` instead of `()` and add a `unsafe .finish_as_vcl_string()` function or whatever.

I think we would need to create a new `WsVclStrBuffer` (with a different name because this one sucks) that filters on input though.

Note that C strings are bad, but UTF could have avoided a whole class of issues by just not allowing null bytes....

**Comment by nyurik** [comment]
we really need to figure out the lifetime aspect - otherwise this wrapper is unsafe by definition and thus - useless.  Lifetime ensures that user is unable to simply store the value in some global structure (e.g. in the VCL context rather than in the context of a single request/workspace) - and return it much later. I do not yet know for certain about the API - may need to think about it a bit more.  What's the actual issue with `WsStr<'ws>`?   Note that you do not need to actually store the `&ws` inside `WsStr` - you can use `PhantomData` for that.

**Comment by gquintard** [comment]
Sorry, I've been busy with other think, but I will look at this very soon™

**Comment by perbu** [comment]
We've discussed this. It is hard.

**Comment by gquintard** [comment]
Alright, I have opinions, but they need to be discussed :-)

* `VCL_STRING`

It's currently `unsafe` to return a `VCL_STRING`, and I think this is the right take. There you go.
We should however have a `VclString<'a>` type that wraps it and that can be only created safely, indexed on the `&'a Ctx` passed to the vmod

* Lack of workspace

I don't reckon this is an issue we should worry about. You ran out of memory, there's almost no good way to recover from this, and even if we do, the next VCL/vmod call that needs the workspace will error out anyway. Not worth our time.

* `&str` or `String` incompatible with `VCL_STRING`

i.e. the returned data contains a `\0`. As it is now, we truncate. It's not great, but it maps exactly onto the C behavior. I admit it's not an excuse, but I feel like it's an easy rule, and people can use blobs if they want binary payload.
Sometimes though, it's better to error out if the data is wrong, so methinks we should have a macro for the return type (maybe also for the arguments?) that tells us if we should truncate, convert lossily or error out.

Thoughts?
